wire-runtime/src/main/java/com/squareup/wire/Message.kt

```diff
@@ -0,0 +1,156 @@
+  @Throws(IOException::class)
+  fun encode(stream: OutputStream) {
+    adapter.encode(stream, this as M)
+  }
+
+  /**
+   * Superclass for protocol buffer message builders.
+   */
+  abstract class Builder<M : Message<M, B>, B : Builder<M, B>> protected constructor() {
+    /**
+     * Caches unknown fields as a [ByteString] when [buildUnknownFields] is called.
+     * When the caller adds an additional unknown field after that, it will be written to the new
+     * [unknownFieldsBuffer] to ensure that all unknown fields are retained between calls to
+     * [buildUnknownFields].
+     */
+    @Transient internal var unknownFieldsByteString = ByteString.EMPTY
+    /**
+     * [Buffer] of the message's unknown fields that is lazily instantiated between calls to
+     * [buildUnknownFields]. It's automatically cleared in [buildUnknownFields], and can also be
+     * manually cleared by calling [clearUnknownFields].
+     */
+    @Transient internal var unknownFieldsBuffer: Buffer? = null
+    @Transient internal var unknownFieldsWriter: ProtoWriter? = null
+
+    fun addUnknownFields(unknownFields: ByteString): Builder<M, B> = apply {
+      if (unknownFields.size > 0) {
+        prepareForNewUnknownFields()
+        unknownFieldsWriter!!.writeBytes(unknownFields)
+      }
+    }
+
+    fun addUnknownField(
+      tag: Int,
+      fieldEncoding: FieldEncoding,
+      value: Any?
+    ): Builder<M, B> = apply {
+      prepareForNewUnknownFields()
+      val protoAdapter = fieldEncoding.rawProtoAdapter() as ProtoAdapter<Any>
+      protoAdapter.encodeWithTag(unknownFieldsWriter!!, tag, value)
+    }
+
+    fun clearUnknownFields(): Builder<M, B> = apply {
+      unknownFieldsByteString = ByteString.EMPTY
+      if (unknownFieldsBuffer != null) {
+        unknownFieldsBuffer!!.clear()
+        unknownFieldsBuffer = null
+      }
+      unknownFieldsWriter = null
+    }
+
+    /**
+     * Returns a byte string with this message's unknown fields. Returns an empty byte string if
+     * this message has no unknown fields.
+     */
+    fun buildUnknownFields(): ByteString {
+      if (unknownFieldsBuffer != null) {
+        // Reads and caches the unknown fields from the buffer.
+        unknownFieldsByteString = unknownFieldsBuffer!!.readByteString()
+        unknownFieldsBuffer = null
+        unknownFieldsWriter = null
+      }
+      return unknownFieldsByteString
+    }
+
+    /** Returns an immutable [Message] based on the fields that set in this builder. */
+    abstract fun build(): M
+
+    private fun prepareForNewUnknownFields() {
+      if (unknownFieldsBuffer == null) {
+        unknownFieldsBuffer = Buffer()
+        unknownFieldsWriter = ProtoWriter(unknownFieldsBuffer!!)
+        // Writes the cached unknown fields to the buffer.
+        unknownFieldsWriter!!.writeBytes(unknownFieldsByteString)
+        unknownFieldsByteString = ByteString.EMPTY
+      }
+    }
+  }
+
+  companion object {
+    private const val serialVersionUID = 0L
```

Comment:
Same question as the other PR. This needs to end up as a static field on the `Message` class. Does it?

Reply:
Yup
![image](https://user-images.githubusercontent.com/1186186/56151764-01133780-5f80-11e9-875c-c6d8faf65598.png)